algorithms/linfa-elasticnet/src/algorithm.rs

```diff
@@ -189,6 +319,92 @@ fn coordinate_descent<'a, F: Float>(
+    x: ArrayView2<'a, F>,
+    y: ArrayView2<'a, F>,
+    tol: F,
+    max_steps: u32,
+    l1_ratio: F,
+    penalty: F,
+) -> (Array2<F>, F, u32) {
+    let n_samples = F::cast(x.shape()[0]);
+    let n_features = x.shape()[1];
+    let n_tasks = y.shape()[1];
+    // the parameters of the model
+    let mut w = Array2::<F>::zeros((n_features, n_tasks));
+    // the residuals: `Y - XW` (since W=0, this is just `Y` for now),
+    // the residuals are updated during the algorithm as the parameters change
+    let mut r = y.to_owned();
+    let mut n_steps = 0u32;
+    let norm_cols_x = x.map_axis(Axis(0), |col| col.dot(&col));
+    let mut gap = F::one() + tol;
+    let d_w_tol = tol;
+    let tol = tol * y.iter().map(|&y_ij| y_ij * y_ij).sum();
+    while n_steps < max_steps {
+        let mut w_max = F::zero();
+        let mut d_w_max = F::zero();
+        for j in 0..n_features {
+            if abs_diff_eq!(norm_cols_x[j], F::zero()) {
+                continue;
+            }
+            let mut old_w_j = w.slice_mut(s![j, ..]);
+            let x_j = x.slice(s![.., j]);
+            let norm_old_w_j = old_w_j.dot(&old_w_j).sqrt();
+            if abs_diff_ne!(norm_old_w_j, F::zero()) {
+                general_mat_mul(
+                    F::one(),
+                    &x_j.view().insert_axis(Axis(1)),
+                    &old_w_j.view().insert_axis(Axis(0)),
+                    F::one(),
+                    &mut r,
+                );
+                //for i in 0..x.shape()[0] {
+                //r.slice_mut(s![i, ..]).scaled_add(x_j[i], &old_w_j);
+                //}
+            }
+            let tmp = x_j.dot(&r);
+            old_w_j.assign(
+                &(block_soft_thresholding(tmp.view(), n_samples * l1_ratio * penalty)
+                    / (norm_cols_x[j] + n_samples * (F::one() - l1_ratio) * penalty)),
+            );
+            let norm_w_j = old_w_j.dot(&old_w_j).sqrt();
+            if abs_diff_ne!(norm_w_j, F::zero()) {
+                general_mat_mul(
+                    -F::one(),
+                    &x_j.insert_axis(Axis(1)),
+                    &old_w_j.insert_axis(Axis(0)),
+                    F::one(),
+                    &mut r,
+                );
+            }
+            let d_w_j = (norm_w_j - norm_old_w_j).abs();
+            d_w_max = F::max(d_w_max, d_w_j);
+            w_max = F::max(w_max, norm_w_j);
+        }
+        n_steps += 1;
+
+        if n_steps == max_steps - 1 || abs_diff_eq!(w_max, F::zero()) || d_w_max / w_max < d_w_tol {
+            // We've hit one potential stopping criteria
+            // check duality gap for ultimate stopping criterion
+            gap = duality_gap_mtl(x.view(), y.view(), w.view(), r.view(), l1_ratio, penalty);
+            if gap < tol {
+                break;
+            }
+        }
+    }
+
+    (w, gap, n_steps)
+}
+
+fn block_soft_thresholding<F: Float>(x: ArrayView1<F>, threshold: F) -> Array1<F> {
```

Comment:
can you add a comment to the block soft thresholding operator (e.g. https://math.stackexchange.com/questions/2045579/deriving-block-soft-threshold-from-l-2-norm-prox-operator) so that people have a reference